substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/InstrumentFeature.java

```diff
@@ -0,0 +1,117 @@
+
+    }
+
+    /**
+     * {@link ReflectionFeature} must come before this feature, because many instrumentation methods
+     * are called by reflection, e.g. premain.
+     */
+    @Override
+    public List<Class<? extends Feature>> getRequiredFeatures() {
+        return List.of(ReflectionFeature.class);
+    }
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        FeatureImpl.AfterRegistrationAccessImpl a = (FeatureImpl.AfterRegistrationAccessImpl) access;
+        cl = a.getImageClassLoader().getClassLoader();
+        ImageSingletons.add(PreMainSupport.class, preMainSupport = new PreMainSupport());
+        if(Options.PremainClasses.hasBeenSet()) {
+            List<String> premains = Options.PremainClasses.getValue().values();
+            for (String premain : premains) {
+                addPremainClass(premain);
+            }
+        }
+    }
+
+
+    /**
+     * Find the premain method from the given class and register it for runtime usage. According
+     * to java.lang.instrument <a
+     * href=https://docs.oracle.com/en/java/javase/17/docs/api/java.instrument/java/lang/instrument/package-summary.html>API
+     * doc</a>, there are two premain methods:
+     * <ol>
+     * <li>{@code public static void premain(String agentArgs, Instrumentation inst)}</li>
+     * <li>{@code public static void premain(String agentArgs)}</li>
+     * </ol>
+     * The first one is taken with higher priority. The second one is taken only when the first
+     * one is absent. <br>
+     * So this method looks for them in the same order.
+     */
+    private void addPremainClass(String premainClass) {
+        try {
+            Class<?> clazz = Class.forName(premainClass, false, cl);
+            Method premain = null;
+            List<Object> args = new ArrayList<>();
+            args.add(""); // First argument is options which will be set at runtime
+            try {
+                premain = clazz.getDeclaredMethod("premain", String.class, Instrumentation.class);
+                args.add(new PreMainSupport.NativeImageNoOpRuntimeInstrumentation());
+            } catch (NoSuchMethodException e) {
+                try {
+                    premain = clazz.getDeclaredMethod("premain", String.class);
+                } catch (NoSuchMethodException e1) {
+                    UserError.abort(e1, "Can't register agent premain method, because can't find the premain method from the given class %s. Please check your -H:%s setting.", premainClass, Options.PremainClasses.getName());
+                }
+            }
+            preMainSupport.registerPremainMethod(premainClass, premain, args.toArray(new Object[0]));
+        } catch (ClassNotFoundException e) {
+            UserError.abort(e,"Can't register agent premain method, because the given class %s is not found. Please check your -H:%s setting.", premainClass, Options.PremainClasses.getName());
```

Comment:
It is better to use `SubstrateOptionsParser.commandArgument` to print the option name. This keeps the correct behavior  when options become public or change the type.